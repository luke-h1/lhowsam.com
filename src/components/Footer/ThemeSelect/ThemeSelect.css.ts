import { variables } from '@frontend/styles/variables.css';
import { style } from '@vanilla-extract/css';
import { calc } from '@vanilla-extract/css-utils';

export const root = style({
  position: 'relative',
  display: 'inline-flex',
  height: '2rem',
  width: 'min-content',
});

export const icon = style({
  position: 'absolute',
  display: 'grid',
  placeItems: 'center',
  left: variables.spacing.sm,
  top: 0,
  width: variables.spacing.md,
  height: '100%',
  pointerEvents: 'none',
});

export const select = style({
  border: '1px solid',
  borderColor: variables.color.border,
  paddingTop: variables.spacing.xs,
  paddingBottom: variables.spacing.xs,
  paddingRight: calc.add(
    calc.multiply(variables.spacing.sm, 2),
    variables.spacing.md,
  ),
  paddingLeft: calc.add(
    calc.multiply(variables.spacing.sm, 2),
    variables.spacing.md,
  ),
  height: '2rem',
  fontSize: variables.fontSize.sm,
  borderRadius: variables.radius.md,
  textAlign: 'left',
  ':focus': {
    outline: 'transparent',
  },
  ':focus-visible': {
    outlineWidth: '2px',
    outlineStyle: 'solid',
    outlineOffset: '2px',
    outlineColor: variables.color.outline,
  },
});

export const arrow = style({
  position: 'absolute',
  display: 'grid',
  placeItems: 'center',
  right: variables.spacing.sm,
  top: 0,
  width: variables.spacing.md,
  height: '100%',
  pointerEvents: 'none',
});
