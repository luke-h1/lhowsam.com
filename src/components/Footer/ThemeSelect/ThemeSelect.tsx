/* eslint-disable jsx-a11y/label-has-associated-control */
import VisuallyHidden from '@frontend/components/VisuallyHidden/VisuallyHidden';
import { useMounted } from '@frontend/hooks/useMounted';
import { useTheme } from 'next-themes';
import { Sun, Moon, Monitor, Code } from 'react-feather';
import * as styles from './ThemeSelect.css';

const icons = {
  system: <Monitor width="0.9em" />,
  light: <Sun width="0.9em" />,
  dark: <Moon width="0.9em" />,
} as const;

type Icon = keyof typeof icons;

const ThemeSelect = () => {
  const { isMounted } = useMounted();
  const { theme, setTheme } = useTheme();

  if (!isMounted) {
    return null;
  }

  return (
    <>
      <VisuallyHidden>
        <label htmlFor="theme-select">Select theme</label>
      </VisuallyHidden>
      <div className={styles.root}>
        <span className={styles.icon}>{icons[theme as Icon]}</span>
        <select
          value={theme}
          onChange={e => setTheme(e.target.value)}
          className={styles.select}
          id="theme-select"
        >
          <option value="system" id="theme-system">
            System
          </option>
          <option value="light" id="theme-light">
            Light
          </option>
          <option value="dark" id="theme-dark">
            Dark
          </option>
        </select>
        <span className={styles.arrow}>
          <Code
            width="0.9em"
            style={{
              transform: 'rotate(90deg)',
            }}
          />
        </span>
      </div>
    </>
  );
};
export default ThemeSelect;
