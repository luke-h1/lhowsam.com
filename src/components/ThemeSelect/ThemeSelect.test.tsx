import render from '@frontend/test/render';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ThemeSelect from '.';

describe('ThemeSelect', () => {
  test('renders correctly', () => {
    render(<ThemeSelect />);
  });
  test('defaults to system theme', () => {
    const { container } = render(<ThemeSelect />);
    expect(screen.getByTestId('ThemeSelect-system')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  test('switches theme to light', async () => {
    render(<ThemeSelect />);

    await userEvent.selectOptions(screen.getByLabelText('Select theme'), [
      screen.getByText('Light'),
    ]);

    expect(screen.getByTestId('ThemeSelect-light')).toBeInTheDocument();
  });
  test('switches theme to dark', async () => {
    render(<ThemeSelect />);

    await userEvent.selectOptions(screen.getByLabelText('Select theme'), [
      screen.getByText('Dark'),
    ]);

    expect(screen.getByTestId('ThemeSelect-dark')).toBeInTheDocument();
  });
});
