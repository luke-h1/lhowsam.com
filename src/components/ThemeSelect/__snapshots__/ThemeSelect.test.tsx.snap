// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSelect defaults to system theme 1`] = `
<div>
  <script>
    !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&true)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}if(e==='light'||e==='dark')d.style.colorScheme=e}catch(e){}}()
  </script>
  <span
    class="VisuallyHidden_root__rd7gei0"
  >
    <label
      for="theme-select"
    >
      Select theme
    </label>
  </span>
  <div
    class="ThemeSelect_root__wmdp110"
  >
    <span
      class="ThemeSelect_icon__wmdp111"
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="0.9em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="14"
          rx="2"
          ry="2"
          width="20"
          x="2"
          y="3"
        />
        <line
          x1="8"
          x2="16"
          y1="21"
          y2="21"
        />
        <line
          x1="12"
          x2="12"
          y1="17"
          y2="21"
        />
      </svg>
    </span>
    <span
      class="VisuallyHidden_root__rd7gei0"
    >
      <label
        for="theme-select"
      >
        Theme
      </label>
    </span>
    <select
      class="ThemeSelect_select__wmdp112"
      id="theme-select"
    >
      <option
        id="theme-system"
        value="system"
      >
        System
      </option>
      <option
        id="theme-light"
        value="light"
      >
        Light
      </option>
      <option
        id="theme-dark"
        value="dark"
      >
        Dark
      </option>
    </select>
    <span
      class="ThemeSelect_arrow__wmdp113"
      data-testid="ThemeSelect-system"
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="transform: rotate(90deg);"
        viewBox="0 0 24 24"
        width="0.9em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="16 18 22 12 16 6"
        />
        <polyline
          points="8 6 2 12 8 18"
        />
      </svg>
    </span>
  </div>
</div>
`;
